### DISCUSSION TITLE
IPython behaviors

### DISCUSSION THREAD

**Discussion by ivandiazzz** [xonsh/xonsh#4239] [Q&A] [answered]
Hello! i find xonsh potentially great for interactive analysis. I currently use IPython and there are two behaviors that I haven't been able to get working on xonsh:

1. In IPython each command is prefixed by a number (In[1], Out[9] for example). While running commands I can get the output of any previous command by typing underscore + command number: _19 for example. This gives me access to the output of any command in my session. Is such a behavior possible with xonsh at the moment?

2. When I type import module+[Tab] in IPython I get module and submodule completions, which are extremely useful. At the moment in my xonsh sessions, pressing the Tab only completes filenames in the present working directory. How can I enable module name completion?

**Comment by anki-code** [comment]
1. First of all try to run `__xonsh__.history[-1]` to get the history record with `out` variable. If it's not enough the behavior you described could be implemented as [xontrib](https://xon.sh/tutorial_xontrib.html) ([xonsh contribution/extension/plugin](https://github.com/topics/xontrib)). The great start is to use [cookiecutter template for the xontrib](https://github.com/xonsh/xontrib-cookiecutter).

2. When I type `from pandas import Da` I've got the suggestion `DataFrame, DateOffset, etc`. Please check that you're using `prompt-toolkit` shell type by running `xonfig` (to install run `python -m pip install 'xonsh[full]'`). Also you can activate the [jedi xontrib](https://xon.sh/xontribs.html#jedi) by running `xontrib load jedi` to get additional suggestions.

**Reply by anki-code** [reply]
@ivandiazzz example around point 1 in `xonsh --no-rc`:
```python
$PROMPT_FIELDS['cmd_num'] = lambda: len(__xonsh__.history)
$PROMPT = $PROMPT + '[{cmd_num}] '
# Now prompt has the number of history record [N]

echo 123  # [5] 
__xonsh__.history[5].out
# '123\n'

# Shortcuts:
out = lambda args: __xonsh__.history[int(args[0]) if type(args) is list else args].out
aliases['out'] = out

out(5)        # Python mode function
# '123\n'
out 5         # Subprocess mode alias
# 123
```

**Reply by ivandiazzz** [reply]
Thank you for your quick answers! I tried playing around with this modification and it is still not giving me the correct behavior. If I simply type an integer and press the Enter key, the "out" is None, for example:

~ $ [2] 1+1
2
~ $ [3] out(2)+1
xonsh: For full traceback set: $XONSH_SHOW_TRACEBACK = True
TypeError: unsupported operand type(s) for +: 'NoneType' and 'int'

whereas I would expect the answer to simply be 3

**Reply by anki-code** [reply]
This looks like a new feature: save the result of python code.

**Reply by dyuri** [reply]
Started to experiment by storing the python return value in history (but not persisted it), you can check it here: https://github.com/dyuri/xonsh/tree/python_return

The return value is stored as `pyrtn`, so anki's code above should be changed to:
```python
out = lambda args: __xonsh__.history[int(args[0]) if type(args) is list else args].pyrtn
```

It's not production ready yet, sometimes breaks, but I'll try to fix it if I have time.

**Reply by dyuri** [reply]
After a bit of struggling I've decided to create a xontrib instead, since this feature might have considerable memory usage impact on xonsh.

**Comment by dyuri** [comment, marked as answer]
I've created a xontrib that provides something similar to `ipython`'s In/Out: https://github.com/dyuri/xontrib-pyrtn

**Reply by daniel-shimon** [reply]
This is awesome!
If you really want to, you can simulate `In[x]` and `Out[x]` by overriding `__getitem__`

**Reply by dyuri** [reply]
Yeah I know, and it's on my TODO list :)

**Reply by dyuri** [reply]
... and done!
